### DISCUSSION TITLE
Are custom rules still disabled by default

### DISCUSSION THREAD

**Discussion by sschuberth** [detekt/detekt#3850] [Q&A]
I'm just wondering whether [the sentence](https://github.com/detekt/detekt/blame/main/docs/pages/extensions.md#L226-L228)

> By default all rules not annotated with `@ActiveByDefault` are disabled.
That means your custom rules are also disabled if you have not explicitly enabled
them in the `detekt` yaml configuration file.

is still true, as my custom rules neither have the `@ActiveByDefault` annotation, nor are they activated in the YAML configuration file, but still they work fine.

So, is this really still valid, or can the sentence be deleted?

**Comment by cortinico** [comment]
> but still they work fine.

What do you mean with _work fine_?

The sentence about `@ActiveByDefault` still holds true. Please note that the annotation is just metadata. Once we decide to enable a rule by default for all the users, we set `enabled: true` for such rule and add the `@ActiveByDefault` annotation.

This means that if the source code of such rule does not contain `@ActiveByDefault`, that rule is not active by default.

Custom rules are completely under your control. So it's up to you to eventually annotate them and enable/disable them as you wish.

**Reply by sschuberth** [reply]
> What do you mean with _work fine_?

I mean that I see violation against the rule being reported although the rule does not have `@ActiveByDefault` nor `enabled: true`.

> Custom rules are completely under your control. So it's up to you to eventually annotate them and enable/disable them as you wish.

Sure. But as explained above the "That means your custom rules are also disabled if you have not explicitly enabled them in the `detekt` yaml configuration file." does not seem to be true for me.

**Reply by cortinico** [reply]
> the rule does not have `@ActiveByDefault` nor `enabled: true`.

Does the rule have `enabled: false`? 
If so: that's a bug.
If not: what is happening is most likely that you have `enabled: true` on the ruleset.
Is this the case?

**Reply by sschuberth** [reply]
> Does the rule have `enabled: false`?

No.

> If not: what is happening is most likely that you have `enabled: true` on the ruleset.

Hmm. *Which* ruleset? How do I actually find out which rulset my custom rule belongs to? [Here is the code](https://github.com/oss-review-toolkit/ort/blob/71b8949a1df5af6995ced653a9af908d225e4ff6/detekt-rules/src/main/kotlin/OrtPackageNaming.kt#L41). Is the ruleset implied by `Severity.Style`? If so, then yes, the `style` ruleset in the YAML seems to be enabled by default.

**Reply by marschwar** [reply]
No, the `ruleSetId` is defined by the `RuleSetProvider`

https://github.com/oss-review-toolkit/ort/blob/71b8949a1df5af6995ced653a9af908d225e4ff6/detekt-rules/src/main/kotlin/OrtRuleSet.kt#L27

Therefor your rule set id is `ORT`. So it is possible to disable all of your rules by adding 
```yaml
ORT:
  active: false
```

to the `.detekt.yml`.

Coming back to your original question. Unless you explicitly configure a rule (set) to be `active: false` , it is considered active.
https://github.com/detekt/detekt/blob/4efa571408a1e4d2c997e86318c9a432bd46972a/detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/rules/RuleSets.kt#L19-L20
So I guess we should update the documentation or change the implementation.

**Reply by marschwar** [reply]
I am afraid, that my answer is not entirely true though. By default rules should be inactive.
https://github.com/detekt/detekt/blob/4efa571408a1e4d2c997e86318c9a432bd46972a/detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/ConfigAware.kt#L44-L48

I will investigate some more.

**Reply by marschwar** [reply]
I think I figured out what is going on. At the moment it is not possible to configure your rules as they do not pass along the provided config. If you change

```kotlin
class OrtImportOrder : Rule() {
```
to
```kotlin
class OrtImportOrder(config: Config) : Rule(config) {
```

it is possible to configure your custom rule
```yaml
ORT:
  active: true
  OrtPackageNaming:
    active: true
  OrtImportOrder:
    active: false
```

So currently you are using the `EmptyConfig` which by default always returns to for the `active` config key.

So the sentence in question holds true after all, even though the documentation should not encourage writing rules that use an `EmptyConfig`.

**Reply by cortinico** [reply]
> So the sentence in question holds true after all, even though the documentation should not encourage writing rules that use an `EmptyConfig`.

Thanks for clarifying @marschwar 🙏  

Fun fact: we could probably create a custom ruleset with rules to validate custom rules code (e.g. `DoNotUseEmptyConfig` and so on).

**Reply by BraisGabin** [reply]
That have a lot of sense. I think that we should go for the template repository. That should help A LOT to have the scaffolding. I'm going to create an issue to track it.
